File: AM_Program/AM_Shared/clsMessageSender.vb
Option Strict On

Imports Apache.NMS

''' <summary>
''' Sends messages to ActiveMQ message broker using NMS client library
''' </summary>
''' <remarks></remarks>
Class clsMessageSender

    Private ReadOnly topicName As String
    Private ReadOnly brokerUri As String
    Private ReadOnly processorName As String

    Private connection As IConnection
    Private session As ISession
    Private producer As IMessageProducer

    Private isDisposed As Boolean = False
    Private hasConnection As Boolean = False

    Public Sub New(brokerUri As String, topicName As String, processorName As String)
        Me.topicName = topicName
        Me.brokerUri = brokerUri
        Me.processorName = processorName
    End Sub

    ''' <summary>
    '''  Send the message using NMS connection objects
    ''' </summary>
    ''' <param name="message"></param>
    ''' <remarks>
    ''' If connection does not exist, make it
    ''' If connection objects don't work, erase them and make another set
    ''' </remarks>
    Public Sub SendMessage(message As String)
        If Me.isDisposed Then
            Exit Sub
        End If
        If Not hasConnection Then
            CreateConnection()
        End If
        If hasConnection Then
            Try
                Dim textMessage As ITextMessage = Me.session.CreateTextMessage(message)
                textMessage.Properties.SetString("ProcessorName", Me.processorName)
                Me.producer.Send(textMessage)
            Catch e As Exception
                ' something went wrong trying to send message,
                ' get rid of current set of connection object - they'll never work again anyway
                DestroyConnection()
                ' temp debug
                '+ e.Message
                '                Console.WriteLine("=== Error sending ===" & Environment.NewLine)
            End Try
        End If
    End Sub

    ''' <summary>
    ''' Create set of NMS connection objects necessary to talk to the ActiveMQ broker
    ''' </summary>
    ''' <param name="retryCount"></param>
    ''' <param name="timeoutSeconds"></param>
    ''' <remarks></remarks>
    Protected Sub CreateConnection(Optional ByVal retryCount As Integer = 2, Optional ByVal timeoutSeconds As Integer = 15)
        If hasConnection Then
            Exit Sub
        End If

        If retryCount < 0 Then
            retryCount = 0
        End If

        Dim retriesRemaining As Integer = retryCount

        If timeoutSeconds < 5 Then
            timeoutSeconds = 5
        End If

        Dim errorList = New List(Of String)()

        While retriesRemaining >= 0
            Try
                Dim connectionFactory As IConnectionFactory = New ActiveMQ.ConnectionFactory(Me.brokerUri)
                Me.connection = connectionFactory.CreateConnection()
                Me.connection.RequestTimeout = New TimeSpan(0, 0, timeoutSeconds)
                Me.connection.Start()

                Me.session = connection.CreateSession()

                Me.producer = Me.session.CreateProducer(New ActiveMQ.Commands.ActiveMQTopic(Me.topicName))
                Me.hasConnection = True

                Return
            Catch ex As Exception
                ' Connection failed
                If Not errorList.Contains(ex.Message) Then
                    errorList.Add(ex.Message)
                End If

                ' Sleep for 3 seconds
                System.Threading.Thread.Sleep(3000)
            End Try

            retriesRemaining -= 1
        End While

        ' If we get here, we never could connect to the message broker

        Dim msg = "Exception creating broker connection"
        If retryCount > 0 Then
            msg += " after " & (retryCount + 1) & " attempts"
        End If

        msg += ": " + String.Join("; ", errorList)
        clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, msg)

        Console.WriteLine("=== Error creating Activemq connection ===" & Environment.NewLine & msg)
   
    End Sub

    Protected Sub DestroyConnection()
        Try
            If hasConnection Then
                If Not Me.producer Is Nothing Then Me.producer.Dispose()
                If Not Me.session Is Nothing Then Me.session.Dispose()
                If Not Me.connection Is Nothing Then Me.connection.Dispose()
                Me.hasConnection = False
            End If
        Catch ex As Exception
            ' Ignore errors here
        End Try
    End Sub

    Public Sub Dispose()
        If Not Me.isDisposed Then
            Me.isDisposed = True
            Try
                Me.DestroyConnection()
            Catch ex As Exception
                ' Ignore errors here
            End Try
        End If
    End Sub
End Class

File: AM_Program/AM_Shared/clsMessageSender.cs

using System;
using System.Collections.Generic;
using Apache.NMS;

namespace AnalysisManagerBase
{

    /// <summary>
    /// Sends messages to ActiveMQ message broker using NMS client library
    /// </summary>
    /// <remarks></remarks>
    class clsMessageSender
    {

        private readonly string topicName;
        private readonly string brokerUri;

        private readonly string processorName;
        private IConnection connection;
        private ISession session;

        private IMessageProducer producer;
        private bool isDisposed;

        private bool hasConnection;

        public clsMessageSender(string brokerUri, string topicName, string processorName)
        {
            this.topicName = topicName;
            this.brokerUri = brokerUri;
            this.processorName = processorName;
        }

        /// <summary>
        ///  Send the message using NMS connection objects
        /// </summary>
        /// <param name="message"></param>
        /// <remarks>
        /// If connection does not exist, make it
        /// If connection objects don't work, erase them and make another set
        /// </remarks>
        public void SendMessage(string message)
        {
            if (isDisposed)
            {
                return;
            }
            if (!hasConnection)
            {
                CreateConnection();
            }
            if (hasConnection)
            {
                try
                {
                    var textMessage = session.CreateTextMessage(message);
                    textMessage.Properties.SetString("ProcessorName", processorName);
                    producer.Send(textMessage);
                }
                catch (Exception)
                {
                    // something went wrong trying to send message,
                    // get rid of current set of connection object - they'll never work again anyway
                    DestroyConnection();                    
                }
            }
        }

        /// <summary>
        /// Create set of NMS connection objects necessary to talk to the ActiveMQ broker
        /// </summary>
        /// <param name="retryCount"></param>
        /// <param name="timeoutSeconds"></param>
        /// <remarks></remarks>
        protected void CreateConnection(int retryCount = 2, int timeoutSeconds = 15)
        {
            if (hasConnection)
            {
                return;
            }

            if (retryCount < 0)
            {
                retryCount = 0;
            }

            var retriesRemaining = retryCount;

            if (timeoutSeconds < 5)
            {
                timeoutSeconds = 5;
            }

            var errorList = new List<string>();

            while (retriesRemaining >= 0)
            {
                try
                {
                    var connectionFactory = new Apache.NMS.ActiveMQ.ConnectionFactory(brokerUri);
                    connection = connectionFactory.CreateConnection();
                    connection.RequestTimeout = new TimeSpan(0, 0, timeoutSeconds);
                    connection.Start();

                    session = connection.CreateSession();

                    producer = session.CreateProducer(new Apache.NMS.ActiveMQ.Commands.ActiveMQTopic(topicName));
                    hasConnection = true;

                    return;
                }
                catch (Exception ex)
                {
                    // Connection failed
                    if (!errorList.Contains(ex.Message))
                    {
                        errorList.Add(ex.Message);
                    }

                    // Sleep for 3 seconds
                    System.Threading.Thread.Sleep(3000);
                }

                retriesRemaining -= 1;
            }

            // If we get here, we never could connect to the message broker

            var msg = "Exception creating broker connection";
            if (retryCount > 0)
            {
                msg += " after " + (retryCount + 1) + " attempts";
            }

            msg += ": " + string.Join("; ", errorList);
            OnErrorEvent(msg);           
        }

        protected void DestroyConnection()
        {
            try
            {
                if (hasConnection)
                {
                    producer?.Dispose();
                    session?.Dispose();
                    connection?.Dispose();
                    hasConnection = false;
                }
            }
            catch (Exception)
            {
                // Ignore errors here
            }
        }

        public void Dispose()
        {
            if (!isDisposed)
            {
                isDisposed = true;
                try
                {
                    DestroyConnection();
                }
                catch (Exception)
                {
                    // Ignore errors here
                }
            }
        }

        #region "Events and Event Handlers"

        public event ErrorEventEventHandler ErrorEvent;
        public delegate void ErrorEventEventHandler(string strMessage, Exception ex);

        /// <summary>
        /// Report an error
        /// </summary>
        /// <param name="strMessage"></param>
        protected void OnErrorEvent(string strMessage)
        {
            ErrorEvent?.Invoke(strMessage, null);
        }

        #endregion
    }
}